File: PEUserMessageHandler/src/main/java/nl/adaptivity/process/userMessageHandler/client/processModel/ProcessNode.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.userMessageHandler.client.processModel;

import java.util.Collection;
import java.util.Map;


public abstract class ProcessNode {

  private static final int HORIZSEP = 100;

  private static final int VERTSEP = 60;

  private final String mId;

  private boolean mPosSet = false;

  private int mX;

  private int mY;

  protected ProcessNode(final String id) {
    mId = id;
  }

  public String getId() {
    return mId;
  }

  public abstract void resolvePredecessors(Map<String, ProcessNode> map);

  public abstract void ensureSuccessor(ProcessNode node);

  public abstract Collection<ProcessNode> getSuccessors();

  public abstract Collection<ProcessNode> getPredecessors();

  public void unsetPos() {
    mPosSet = false;
  }

  public boolean hasPos() {
    return mPosSet;
  }

  public int layout(final int x, final int y, final ProcessNode source, final boolean forward) {
    if (mPosSet) {
      boolean dx = false;
      boolean dy = false;
      if (x != mX) {
        if (forward) {
          if (x > mX) {
            mX = x;
            dx = true;
          } else { // pX < mX
            mX -= (mX - x) / 2; // center
          }
        } else {
          if (x < mX) {
            mX = x;
            dx = true;
          } else { // pX > mX
            mX += (x - mX) / 2; // center
          }
        }
      }
      if (y != mY) {
        if (y > mY) {
          mY = y;
          dy = true;
        } else {

        }
      }
      if (dx || dy) {
        if (forward) {
          return Math.max(mY, layoutSuccessors(this));
        } else {
          return Math.max(layoutPredecessors(this), mY);
        }
      }
      return mY;

    } else {
      mPosSet = true;
      mX = x;
      if (forward) {
        final int cnt = getPredecessors().size();
        int index = -1;
        int i = 0;
        for (final ProcessNode n : getPredecessors()) {
          if (n == source) {
            index = i;
            break;
          }
          ++i;
        }
        if (index >= 0) {
          mY = (y - ((index * VERTSEP))) + (((cnt - 1) * VERTSEP) / 2);
        } else {
          mY = y;
        }
      } else {
        mY = y;
      }
      return Math.max(layoutPredecessors(source), layoutSuccessors(source));
    }
  }

  private int layoutSuccessors(final ProcessNode source) {
    final Collection<ProcessNode> successors = getSuccessors();
    int posY = mY - (((successors.size() - 1) * VERTSEP) / 2);
    final int posX = mX + HORIZSEP;

    for (final ProcessNode successor : successors) {
      if (successor != source) {
        successor.layout(posX, posY, this, true);
      }
      posY += VERTSEP;
    }
    return Math.min(mY, posY - VERTSEP);
  }

  private int layoutPredecessors(final ProcessNode source) {
    final Collection<ProcessNode> predecessors = getPredecessors();
    int posY = mY - (((predecessors.size() - 1) * VERTSEP) / 2);
    final int posX = mX - HORIZSEP;

    for (final ProcessNode predecessor : predecessors) {
      if (predecessor != source) {
        predecessor.layout(posX, posY, this, false);
      }
      posY += VERTSEP;
    }
    return Math.min(mY, posY - VERTSEP);
  }

  public int getX() {
    return mX;
  }

  public int getY() {
    return mY;
  }

  public void offset(final int offsetX, final int offsetY) {
    mX += offsetX;
    mY += offsetY;
  }

}

File: PE-common/src/commonMain/kotlin/nl/adaptivity/process/processModel/ProcessNode.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.process.processModel

import net.devrieze.util.collection.replaceBy
import nl.adaptivity.diagram.Positioned
import nl.adaptivity.process.util.Identifiable
import nl.adaptivity.process.util.Identified
import nl.adaptivity.process.util.Identifier
import nl.adaptivity.process.util.IdentifyableSet


/**
 * Created by pdvrieze on 27/11/16.
 */
interface ProcessNode : Positioned, Identifiable {

    val ownerModel: ProcessModel<ProcessNode>?

    val predecessors: IdentifyableSet<Identified>

    val successors: IdentifyableSet<Identified>

    val maxSuccessorCount: Int

    val maxPredecessorCount: Int

    val label: String?

    val results: List<IXmlResultType>

    val defines: List<IXmlDefineType>

    val idBase: String

    val isMultiInstance: Boolean

    fun builder(): Builder

    fun asT(): ProcessNode

    fun isPredecessorOf(node: ProcessNode): Boolean

    fun <R> visit(visitor: Visitor<R>): R

    fun getResult(name: String): IXmlResultType?

    fun getDefine(name: String): IXmlDefineType?

    @ProcessModelDSL
    interface Builder {
        val predecessors: Set<Identified>
        val successors: Set<Identified>
        var id: String?
        var label: String?
        var x: Double
        var y: Double
        val defines: MutableCollection<IXmlDefineType>
        val results: MutableCollection<IXmlResultType>
        val idBase: String
        var isMultiInstance: Boolean

        fun result(builder: XmlResultType.Builder.() -> Unit) {
            results.add(XmlResultType.Builder().apply(builder).build())
        }

        fun <R> visit(visitor: BuilderVisitor<R>): R

        fun setDefines(value: Iterable<IXmlDefineType>) = defines.replaceBy(value)
        fun setResults(value: Iterable<IXmlResultType>) = results.replaceBy(value)

        /** Add a successor in an appropriate way for the kind. It may fail if not valid */
        fun addSuccessor(identifier: Identifier)

        /** Add a predecessor in an appropriate way for the kind. It may fail if not valid */
        fun addPredecessor(identifier: Identifier)

        /** Remove a successor in an appropriate way for the kind. It may fail if not valid */
        fun removeSuccessor(identifier: Identifiable)

        /** Remove a predecessor in an appropriate way for the kind. It may fail if not valid */
        fun removePredecessor(identifier: Identifiable)

        fun getResult(name: String): IXmlResultType? = results.firstOrNull { it.name == name }

        fun getDefine(name: String): IXmlDefineType? = defines.firstOrNull { it.name == name }

    }

    interface BuilderVisitor<R> {
        fun visitStartNode(startNode: StartNode.Builder): R
        fun visitEventNode(eventNode: EventNode.Builder): R
        fun visitMessageActivity(activity: MessageActivity.Builder): R = visitGenericActivity(activity)

        fun visitCompositeActivity(activity: CompositeActivity.ModelBuilder): R = visitGenericActivity(activity)

        fun visitReferenceActivity(activity: CompositeActivity.ReferenceBuilder): R = visitGenericActivity(activity)

        fun visitSplit(split: Split.Builder): R
        fun visitJoin(join: Join.Builder): R
        fun visitEndNode(endNode: EndNode.Builder): R
        fun visitGenericActivity(builder: Activity.Builder): R {
            throw UnsupportedOperationException("This visitor does not support handling generic activities")
        }
    }

    interface Visitor<R> {
        fun visitStartNode(startNode: StartNode): R = visitGeneralNode(startNode)
        fun visitEventNode(eventNode: EventNode): R = visitGeneralNode(eventNode)
        fun visitActivity(messageActivity: MessageActivity): R = visitGenericActivity(messageActivity)
        fun visitCompositeActivity(compositeActivity: CompositeActivity): R = visitGenericActivity(compositeActivity)

        fun visitGenericActivity(activity: Activity): R = visitGeneralNode(activity)

        fun visitSplit(split: Split): R = visitGeneralNode(split)
        fun visitJoin(join: Join): R = visitGeneralNode(join)
        fun visitEndNode(endNode: EndNode): R = visitGeneralNode(endNode)
        fun visitGeneralNode(node: ProcessNode): R {
            error("Unsupported node type: ${node}")
        }
    }
}


inline operator fun StartNode.Builder?.invoke(body: StartNode.Builder.() -> Unit) {
    this?.body()
}

inline operator fun MessageActivity.Builder?.invoke(body: MessageActivity.Builder.() -> Unit) {
    this?.body()
}

inline operator fun Split.Builder?.invoke(body: Split.Builder.() -> Unit) {
    this?.body()
}

inline operator fun Join.Builder?.invoke(body: Join.Builder.() -> Unit) {
    this?.body()
}

inline operator fun EndNode.Builder?.invoke(body: EndNode.Builder.() -> Unit) {
    this?.body()
}

internal inline fun ProcessNode.Builder.removeAllPredecessors(predicate: (Identified) -> Boolean) {
    predecessors.filter(predicate).forEach { removePredecessor(it.identifier) }
}

internal inline fun ProcessNode.Builder.removeAllSuccessors(predicate: (Identified) -> Boolean) {
    successors.filter(predicate).forEach { removeSuccessor(it.identifier) }
}
